tools/zarr-cache-stats.py

```diff
@@ -0,0 +1,299 @@
+                    r"(?P<timestamp>\d{4}-\d\d-\d\dT\d\d:\d\d:\d\d\.\d+[-+]\d\d:\d\d)"
+                    r" (?P<target>\S+): ",
+                    line,
+                )
+                if not m:
+                    log.warning("Failed to parse Heroku log line: %s", line)
+                    continue
+                timestamp = datetime.fromisoformat(m["timestamp"])
+                target = m["target"]
+                message = line[m.end() :].strip()
+                yield HerokuLog(
+                    timestamp=timestamp, target=target, message=message, line=line
+                )
+
+
+def process_logs(logfiles: Iterable[Path]) -> Events:
+    events = Events()
+    misses_in_progress = {}
+    last_accesses = {}
+    for lg in iterlogs(logfiles):
+        if lg.target == "app[web.1]":
+            entry = json.loads(lg.message)
+            timestamp = datetime.fromisoformat(entry["timestamp"])
+            fields = entry.get("fields", {})
+            if "cache_event" in fields and fields.get("cache") == "zarr-manifests":
+                span_id: str | None = entry.get("span", {}).get("id")
+                manifest_path: str = fields["manifest"]
+                match fields["cache_event"]:
+                    case "miss_pre":
+                        assert span_id is not None
+                        misses_in_progress[
+                            MissKey(span_id=span_id, manifest_path=manifest_path)
+                        ] = MissMoment(
+                            timestamp=timestamp,
+                            approx_cache_len=fields["approx_cache_len"],
+                        )
+                    case "miss_post":
+                        assert span_id is not None
+                        key = MissKey(span_id=span_id, manifest_path=manifest_path)
+                        if (start := misses_in_progress.pop(key, None)) is not None:
+                            end = MissMoment(
+                                timestamp=timestamp,
+                                approx_cache_len=fields["approx_cache_len"],
+                            )
+                            events.misses.append(
+                                Miss(
+                                    manifest_path=manifest_path,
+                                    start=start,
+                                    end=end,
+                                    parallel=len(misses_in_progress),
+                                )
+                            )
+                        last_accesses[manifest_path] = timestamp
+                    case "hit":
+                        last_access = last_accesses.get(manifest_path)
+                        if last_access is not None:
+                            recency_index = sum(
+                                1 for ts in last_accesses.values() if ts > last_access
+                            )
+                        else:
+                            recency_index = len(last_accesses)
```

Comment:
when/how do we hit this on `"hit"`?... not exactly following

Reply:
This happens if the provided logs start in the middle of a server process and a hit is encountered for an entry that was previously accessed before the start of the logs.